File: 450-delete-node-in-a-bst/450-delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        
        if(root==NULL) return root;
        
        if(root->val==key) return helper(root); // if root is to be deleted
        
        TreeNode* curr = root;
        
        // search key in bst
        while(root){
            if(root->val>key){ // in left half;
                if(root->left!=NULL and root->left->val==key){ // check if left node is our key
                    root->left = helper(root->left); // if yes delete
                    break;
                }else{ // if no search further 
                    root = root->left;
                }
            }else{ // in right half;
                if(root->right!=NULL and root->right->val==key){
                    root->right = helper(root->right);
                    break;
                }else{
                    root = root->right;
                }
            }
        }
        
        return curr;
        
        
    }
    
    TreeNode* helper(TreeNode* root){
        if(root->left==NULL) return root->right;
        if(root->right==NULL) return root->left;
        
        TreeNode* rightchild = root->right;
        TreeNode* lastRightInLeft = findLastRight(root->left);
        lastRightInLeft->right  = rightchild;
        return root->left;
        
    }
    
    TreeNode* findLastRight(TreeNode* root){
        if(root->right == NULL){
            return root;
        }
        return findLastRight(root->right);
    }
    
};

File: 450-delete-node-in-a-bst/450-delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    # to merge tree after deletion
    def helper(self,root):
        
        if root.left is None:
            return root.right
        
        elif root.right is None:
            return root.left
        
        rightNode = root.right
        lastRight = self.findLastRight(root.left)
        lastRight.right = rightNode
        return root.left
    
    
    ## get last right
    def findLastRight(self,root):
        
        if root.right is None:
            return root
        
        return self.findLastRight(root.right)
    
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        
        if root is None:
            return root
        
        if root.val == key:
            return self.helper(root)
        
        dummy = root
        
        while root:
            
            if root.val > key: ## go to left
                
                if root.left != None and root.left.val == key:
                    root.left = self.helper(root.left)
                    break
                else:
                    root = root.left
            
            else: #goto right
                
                if root.right !=None and root.right.val == key:
                    root.right = self.helper(root.right)
                    break
                else:
                    root = root.right
                
        return dummy

        